### DISCUSSION TITLE
Add Title Generator to Support Programmatic Titles

### DISCUSSION THREAD

**Discussion by zkulis** [pydantic/pydantic#4044] [Feature Request]
## Background
Currently, Pydantic allows specifying a string literal for the `title` value used in the exported JSON Schema. For `BaseModel`, this is accomplished via the `Config` class:

```
from pydantic import BaseModel

class MyDataClass(BaseModel):
   class Config:
      title = "My Data Class"
```

For `BaseModel` fields, an explicit `title` string may be specified with a `Field` instance:
```
from pydantic import BaseModel, Field
from typing import List

class MyDataClass(BaseModel):
   matchCriteria: List[str] = Field(default_factory=list, title="Match Criteria")
   class Config:
      title = "My Data Class"
```

## Feature Request
It would be beneficial if Pydantic provided a programmatic way to set the `title` value for both models and fields, similar to the functionality provided by `alias_generator` [here](https://pydantic-docs.helpmanual.io/usage/model_config/#alias-generator). Specifically:

1. A `title_generator` class attribute in the `Config` class, and
2. A `title_generator` parameter in the `Field` constructor. 

In both cases, the `title_generator` would take a Python callable that receives:
1. The `BaseModel` class name for the first case
2. The field alias for the second case

This is demonstrated below:

```
import re
from pydantic import BaseModel

def make_title(name: str):
   def _capitalize(v: str):
      return v[0].upper() + v[1:]
   return re.sub(r"(?<=[a-z])([A-Z])", r" \1", _capitalize(name))

class MyDataClass(BaseModel):
   matchCriteria: List[str] = Field(default_factory=list, title_generator=make_title)
   class Config:
      title_generator=make_title
```

The resulting JSON Schema would be:
```
import json
print(json.dumps(MyDataClass.schema(), indent=2))

{
  "title": "My Data Class",
  "type": "object",
  "properties": {
    "matchCriteria": {
      "title": "Match Criteria",
      "type": "array",
      "items": {
        "type": "string"
      }
    }
  }
}
```

**Comment by Galarzaa90** [comment]
I was about to open a new issue for this, but I'm glad I found this before.

I second this, this would be very useful. Having to add a title to every field made up of multiple words is annoying, but the default title looks awful.

![image](https://user-images.githubusercontent.com/12865379/196522043-685520b4-6181-4cfa-9760-5037d5eeb43c.png)

In fact, just being able to not generate a title for them at all would be useful too, but setting `title=None` has no effect.

**Reply by samuelcolvin** [reply]
Agreed, I've created #4632 to track this as a feature request.
